### PR TITLE
extend network support for HermitCore

### PR DESCRIPTION
**Pull request by stlankes**
- add basic support of TcpListerner for HermitCore
- revise TcpStream to support peer_addr

### CODE REVIEW

File: src/libstd/sys/hermit/net.rs [hunk 1]
```diff
@@ -125,42 +141,64 @@ impl TcpStream {
...
+            Ipv6(ref addr) => SocketAddr::new(
+                IpAddr::V6(Ipv6Addr::new(
+                    ((addr.0[0] as u16) << 8) | addr.0[1] as u16,
+                    ((addr.0[2] as u16) << 8) | addr.0[3] as u16,
+                    ((addr.0[4] as u16) << 8) | addr.0[5] as u16,
+                    ((addr.0[6] as u16) << 8) | addr.0[7] as u16,
+                    ((addr.0[8] as u16) << 8) | addr.0[9] as u16,
+                    ((addr.0[10] as u16) << 8) | addr.0[11] as u16,
+                    ((addr.0[12] as u16) << 8) | addr.0[13] as u16,
+                    ((addr.0[14] as u16) << 8) | addr.0[15] as u16)),
```
Comment (lzutao) [inline src/libstd/sys/hermit/net.rs:161]:
```suggestion
                IpAddr::V6(Ipv6Addr::from(addr.0),
```

File: src/libstd/sys/hermit/net.rs [hunk 2]
```diff
@@ -169,40 +207,61 @@ impl TcpStream {
...
+        let (handle, ipaddr, port) = abi::tcplistener::accept(self.0.port())
+            .map_err(|_| io::Error::new(ErrorKind::Other, "accept failed"))?;
+        let saddr = match ipaddr {
+            Ipv4(ref addr) => SocketAddr::new(
+                IpAddr::V4(Ipv4Addr::new(addr.0[0], addr.0[1], addr.0[2], addr.0[3])),
+                port,
+            ),
+            Ipv6(ref addr) => SocketAddr::new(
+                IpAddr::V6(Ipv6Addr::new(
+                    ((addr.0[0] as u16) << 8) | addr.0[1] as u16,
```
Comment (lzutao) [inline src/libstd/sys/hermit/net.rs:245]: simiar

File: src/libstd/sys/hermit/net.rs [hunk 3]
```diff
@@ -17,14 +20,31 @@ pub fn init() -> io::Result<()> {
...
+
+impl Drop for Socket {
+    fn drop(&mut self) {
+        let _ = abi::tcpstream::close(self.0);
+    }
+}
+
+
+#[derive(Clone)]
+pub struct TcpStream(Arc<Socket>);
```
Comment (kennytm) [inline src/libstd/sys/hermit/net.rs:40]: why we need the `Arc` here?

Comment (stlankes) [reply src/libstd/sys/hermit/net.rs:39]: I want to close it once. Hm, should I use a reference to the handle?

File: src/libstd/sys/hermit/net.rs (hunk 1 above)
Comment (stlankes) [reply src/libstd/sys/hermit/net.rs]: This doesn't work on my system. How do you convert an array of u8 to an army of u16?

Comment (lzutao) [reply src/libstd/sys/hermit/net.rs:160]: There is `impl From<[u8; 16]> for Ipv6Addr`: https://doc.rust-lang.org/nightly/std/net/struct.Ipv6Addr.html#impl-From%3C%5Bu8%3B%2016%5D%3E

Comment (stlankes) [reply src/libstd/sys/hermit/net.rs:160]: Ok, I will test it. Thanks for the hint!

File: src/libstd/sys/hermit/net.rs (hunk 3 above)
Comment (kennytm) [reply src/libstd/sys/hermit/net.rs:39]: I see. Could you add a comment in the code explaining the use of Arc?

Comment (stlankes) [reply src/libstd/sys/hermit/net.rs:39]: Yeah, good idea!

File: src/libstd/sys/hermit/net.rs [hunk 4]
```diff
@@ -125,42 +143,53 @@ impl TcpStream {
     }
 
     pub fn peer_addr(&self) -> io::Result<SocketAddr> {
-        Err(io::Error::new(ErrorKind::Other, "peer_addr isn't supported"))
+        let (ipaddr, port) = abi::tcpstream::peer_addr(*self.0.as_inner())
+            .map_err(|_| io::Error::new(ErrorKind::Other, "peer_addr failed"))?;
+
+        let saddr = match ipaddr {
+            Ipv4(ref addr) => SocketAddr::new(
+                IpAddr::V4(Ipv4Addr::new(addr.0[0], addr.0[1], addr.0[2], addr.0[3])),
```
Comment (lzutao) [inline src/libstd/sys/hermit/net.rs:151]:
```suggestion
                IpAddr::V4(Ipv4Addr::from(addr.0)),
```

File: src/libstd/sys/hermit/net.rs [hunk 5]
```diff
@@ -169,40 +198,50 @@ impl TcpStream {
...
+        Ok(self.0)
     }
 
     pub fn accept(&self) -> io::Result<(TcpStream, SocketAddr)> {
-        Err(io::Error::new(ErrorKind::Other, "not supported"))
+        let (handle, ipaddr, port) = abi::tcplistener::accept(self.0.port())
+            .map_err(|_| io::Error::new(ErrorKind::Other, "accept failed"))?;
+        let saddr = match ipaddr {
+            Ipv4(ref addr) => SocketAddr::new(
+                IpAddr::V4(Ipv4Addr::new(addr.0[0], addr.0[1], addr.0[2], addr.0[3])),
```
Comment (lzutao) [inline src/libstd/sys/hermit/net.rs:231]:
```suggestion
                IpAddr::V4(Ipv4Addr::from(addr.0)),
```

File: src/libstd/sys/hermit/net.rs (hunk 4 above)
Comment (lzutao) [reply src/libstd/sys/hermit/net.rs:151]: Resolving in #73389 .

**Pull request merged**